<?xml version="1.0"?>

<robot name="stewart">
  
  <!-- *************************************************** -->
  <!-- ******************** Kinematics ******************* -->
  <!-- *************************************************** -->
  <!-- Note the 'origin' tag in the joints defines the
       shift and reorientation before the continuous joint
       rotates around the 'axis'. -->

  <link name="world">
    <visual>
      <origin rpy="3.14159 0 0" xyz="0 0 0"/>
      <geometry>
        <mesh filename="package://finalprojectcode/meshes/Platform_Top.dae"/>
      </geometry>
    </visual>
  </link>

# LEG 1
  <joint name="leg1-pitch" type="continuous">
    <parent link="world"/>
    <child  link="leg1-uj-shaft"/>
    <origin xyz="0 1.25 -0.2" rpy="0 0 0"/>
    <axis   xyz="1 0 0"/>
  </joint>

  <link name="leg1-uj-shaft">
    <visual>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <geometry>
        <mesh filename="package://finalprojectcode/meshes/UJ_Shaft.dae"/>
      </geometry>
    </visual>
  </link>


  <joint name="leg1-roll" type="continuous">
    <parent link="leg1-uj-shaft"/>
    <child  link="leg1-slider-base"/>
    <origin xyz="0 0 0" rpy="0 0 0"/>
    <axis   xyz="0 1 0"/>
  </joint>

  <link name="leg1-slider-base">
    <visual>
      <origin rpy="3.14159 0 0" xyz="0 0 0"/>
      <geometry>
        <mesh filename="package://finalprojectcode/meshes/Slider_Base.dae"/>
      </geometry>
    </visual>
  </link>

  <joint name="leg1-prismatic" type="prismatic">
    <parent link="leg1-slider-base"/>
    <child  link="leg1-slider-shaft"/>
    <origin xyz="0 0 -0.4" rpy="0 0 0"/>
    <limit effort="1000.0" lower="0" upper="1.5" velocity="0.5"/>
    <axis   xyz="0 0 -1"/>
  </joint>

  <link name="leg1-slider-shaft">
    <visual>
      <origin rpy="3.14159 0 0" xyz="0 0 0"/>
      <geometry>
        <mesh filename="package://finalprojectcode/meshes/Slider_Shaft.dae"/>
      </geometry>
    </visual>
  </link>

  <joint name="leg1-to-tip" type="fixed">
    <parent link="leg1-slider-shaft"/>
    <child  link="leg1-tip"/>
    <origin xyz="0 0 0" rpy="0 0 0"/>
  </joint>

  <link name="leg1-tip"/>

##### LEG 2 #####

# -sin 60 sin 60 = x y
# rotate 60 deg for joint
  <joint name="leg2-pitch" type="continuous">
    <parent link="world"/>
    <child  link="leg2-uj-shaft"/>
    <origin xyz="-1.08253 0.625 -0.2" rpy="0 0 1.04719"/>
    <axis   xyz="1 0 0"/>
  </joint>

  <link name="leg2-uj-shaft">
    <visual>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <geometry>
        <mesh filename="package://finalprojectcode/meshes/UJ_Shaft.dae"/>
      </geometry>
    </visual>
  </link>


  <joint name="leg2-roll" type="continuous">
    <parent link="leg2-uj-shaft"/>
    <child  link="leg2-slider-base"/>
    <origin xyz="0 0 0" rpy="0 0 0"/>
    <axis   xyz="0 1 0"/>
  </joint>

  <link name="leg2-slider-base">
    <visual>
      <origin rpy="3.14159 0 0" xyz="0 0 0"/>
      <geometry>
        <mesh filename="package://finalprojectcode/meshes/Slider_Base.dae"/>
      </geometry>
    </visual>
  </link>

  <joint name="leg2-prismatic" type="prismatic">
    <parent link="leg2-slider-base"/>
    <child  link="leg2-slider-shaft"/>
    <origin xyz="0 0 -0.4" rpy="0 0 0"/>
    <limit effort="1000.0" lower="0" upper="1.5" velocity="0.5"/>
    <axis   xyz="0 0 -1"/>
  </joint>

  <link name="leg2-slider-shaft">
    <visual>
      <origin rpy="3.14159 0 0" xyz="0 0 0"/>
      <geometry>
        <mesh filename="package://finalprojectcode/meshes/Slider_Shaft.dae"/>
      </geometry>
    </visual>
  </link>

  <joint name="leg2-to-tip" type="fixed">
    <parent link="leg2-slider-shaft"/>
    <child  link="leg2-tip"/>
    <origin xyz="0 0 0" rpy="0 0 0"/>
  </joint>

  <link name="leg2-tip"/>
  
  ##### LEG 3
  
  # cos(240) sin (240) = y
  # rotate 120
  <joint name="leg3-pitch" type="continuous">
    <parent link="world"/>
    <child  link="leg3-uj-shaft"/>
    <origin xyz="-1.08253 -0.625 -0.2" rpy="0 0 2.094395"/>
    <axis   xyz="1 0 0"/>
  </joint>

  <link name="leg3-uj-shaft">
    <visual>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <geometry>
        <mesh filename="package://finalprojectcode/meshes/UJ_Shaft.dae"/>
      </geometry>
    </visual>
  </link>


  <joint name="leg3-roll" type="continuous">
    <parent link="leg3-uj-shaft"/>
    <child  link="leg3-slider-base"/>
    <origin xyz="0 0 0" rpy="0 0 0"/>
    <axis   xyz="0 1 0"/>
  </joint>

  <link name="leg3-slider-base">
    <visual>
      <origin rpy="3.14159 0 0" xyz="0 0 0"/>
      <geometry>
        <mesh filename="package://finalprojectcode/meshes/Slider_Base.dae"/>
      </geometry>
    </visual>
  </link>

  <joint name="leg3-prismatic" type="prismatic">
    <parent link="leg3-slider-base"/>
    <child  link="leg3-slider-shaft"/>
    <origin xyz="0 0 -0.4" rpy="0 0 0"/>
    <limit effort="1000.0" lower="0" upper="1.5" velocity="0.5"/>
    <axis   xyz="0 0 -1"/>
  </joint>

  <link name="leg3-slider-shaft">
    <visual>
      <origin rpy="3.14159 0 0" xyz="0 0 0"/>
      <geometry>
        <mesh filename="package://finalprojectcode/meshes/Slider_Shaft.dae"/>
      </geometry>
    </visual>
  </link>

  <joint name="leg3-to-tip" type="fixed">
    <parent link="leg3-slider-shaft"/>
    <child  link="leg3-tip"/>
    <origin xyz="0 0 0" rpy="0 0 0"/>
  </joint>

  <link name="leg3-tip"/>
  
  ##### LEG 4
  
  # 1.25sin(180) 1.25cos (180) = y
  # rotate 0 / 180
  <joint name="leg4-pitch" type="continuous">
    <parent link="world"/>
    <child  link="leg4-uj-shaft"/>
    <origin xyz="0 -1.25 -0.2" rpy="0 0 0"/>
    <axis   xyz="1 0 0"/>
  </joint>

  <link name="leg4-uj-shaft">
    <visual>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <geometry>
        <mesh filename="package://finalprojectcode/meshes/UJ_Shaft.dae"/>
      </geometry>
    </visual>
  </link>


  <joint name="leg4-roll" type="continuous">
    <parent link="leg4-uj-shaft"/>
    <child  link="leg4-slider-base"/>
    <origin xyz="0 0 0" rpy="0 0 0"/>
    <axis   xyz="0 1 0"/>
  </joint>

  <link name="leg4-slider-base">
    <visual>
      <origin rpy="3.14159 0 0" xyz="0 0 0"/>
      <geometry>
        <mesh filename="package://finalprojectcode/meshes/Slider_Base.dae"/>
      </geometry>
    </visual>
  </link>

  <joint name="leg4-prismatic" type="prismatic">
    <parent link="leg4-slider-base"/>
    <child  link="leg4-slider-shaft"/>
    <origin xyz="0 0 -0.4" rpy="0 0 0"/>
    <limit effort="1000.0" lower="0" upper="1.5" velocity="0.5"/>
    <axis   xyz="0 0 -1"/>
  </joint>

  <link name="leg4-slider-shaft">
    <visual>
      <origin rpy="3.14159 0 0" xyz="0 0 0"/>
      <geometry>
        <mesh filename="package://finalprojectcode/meshes/Slider_Shaft.dae"/>
      </geometry>
    </visual>
  </link>

  <joint name="leg4-to-tip" type="fixed">
    <parent link="leg4-slider-shaft"/>
    <child  link="leg4-tip"/>
    <origin xyz="0 0 0" rpy="0 0 0"/>
  </joint>

  <link name="leg4-tip"/>

  ##### LEG 5
  
  # 1.25sin(120) 1.25cos (120) = y
  # rotate 60
  <joint name="leg5-pitch" type="continuous">
    <parent link="world"/>
    <child  link="leg5-uj-shaft"/>
    <origin xyz="1.0825317 -0.625 -0.2" rpy="0 0 1.047197"/>
    <axis   xyz="1 0 0"/>
  </joint>

  <link name="leg5-uj-shaft">
    <visual>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <geometry>
        <mesh filename="package://finalprojectcode/meshes/UJ_Shaft.dae"/>
      </geometry>
    </visual>
  </link>


  <joint name="leg5-roll" type="continuous">
    <parent link="leg5-uj-shaft"/>
    <child  link="leg5-slider-base"/>
    <origin xyz="0 0 0" rpy="0 0 0"/>
    <axis   xyz="0 1 0"/>
  </joint>

  <link name="leg5-slider-base">
    <visual>
      <origin rpy="3.14159 0 0" xyz="0 0 0"/>
      <geometry>
        <mesh filename="package://finalprojectcode/meshes/Slider_Base.dae"/>
      </geometry>
    </visual>
  </link>

  <joint name="leg5-prismatic" type="prismatic">
    <parent link="leg5-slider-base"/>
    <child  link="leg5-slider-shaft"/>
    <origin xyz="0 0 -0.4" rpy="0 0 0"/>
    <limit effort="1000.0" lower="0" upper="1.5" velocity="0.5"/>
    <axis   xyz="0 0 -1"/>
  </joint>

  <link name="leg5-slider-shaft">
    <visual>
      <origin rpy="3.14159 0 0" xyz="0 0 0"/>
      <geometry>
        <mesh filename="package://finalprojectcode/meshes/Slider_Shaft.dae"/>
      </geometry>
    </visual>
  </link>

  <joint name="leg5-to-tip" type="fixed">
    <parent link="leg5-slider-shaft"/>
    <child  link="leg5-tip"/>
    <origin xyz="0 0 0" rpy="0 0 0"/>
  </joint>

  <link name="leg5-tip"/>
  
  
  ##### LEG 6
  
  # 1.25sin(60) 1.25cos (60) = y
  # rotate -60
  <joint name="leg6-pitch" type="continuous">
    <parent link="world"/>
    <child  link="leg6-uj-shaft"/>
    <origin xyz="1.0825317 0.625 -0.2" rpy="0 0 -1.047197"/>
    <axis   xyz="1 0 0"/>
  </joint>

  <link name="leg6-uj-shaft">
    <visual>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <geometry>
        <mesh filename="package://finalprojectcode/meshes/UJ_Shaft.dae"/>
      </geometry>
    </visual>
  </link>


  <joint name="leg6-roll" type="continuous">
    <parent link="leg6-uj-shaft"/>
    <child  link="leg6-slider-base"/>
    <origin xyz="0 0 0" rpy="0 0 0"/>
    <axis   xyz="0 1 0"/>
  </joint>

  <link name="leg6-slider-base">
    <visual>
      <origin rpy="3.14159 0 0" xyz="0 0 0"/>
      <geometry>
        <mesh filename="package://finalprojectcode/meshes/Slider_Base.dae"/>
      </geometry>
    </visual>
  </link>

  <joint name="leg6-prismatic" type="prismatic">
    <parent link="leg6-slider-base"/>
    <child  link="leg6-slider-shaft"/>
    <origin xyz="0 0 -0.4" rpy="0 0 0"/>
    <limit effort="1000.0" lower="0" upper="1.5" velocity="0.5"/>
    <axis   xyz="0 0 -1"/>
  </joint>

  <link name="leg6-slider-shaft">
    <visual>
      <origin rpy="3.14159 0 0" xyz="0 0 0"/>
      <geometry>
        <mesh filename="package://finalprojectcode/meshes/Slider_Shaft.dae"/>
      </geometry>
    </visual>
  </link>

  <joint name="leg6-to-tip" type="fixed">
    <parent link="leg6-slider-shaft"/>
    <child  link="leg6-tip"/>
    <origin xyz="0 0 0" rpy="0 0 0"/>
  </joint>

  <link name="leg6-tip"/>
  


  <!-- *************************************************** -->
  <!-- ********************* Graphics ******************** -->
  <!-- *************************************************** -->
  <!-- The following is purely to make this look roboticy. -->
  <!-- The 'origin' tag in the link/visual defines the
       center of the object (box, cylinder, sphere, mesh)
       but does not affect the joint chain kinematics. -->

  <!-- *************** Materials / Colors **************** -->

  <material name="white">    <color rgba="1.00 1.00 1.00 1"/>  </material>
  <material name="gray">     <color rgba="0.60 0.60 0.60 1"/>  </material>
  <material name="table">    <color rgba="0.85 0.77 0.77 1"/>  </material>
  <material name="black">    <color rgba="0.00 0.00 0.00 1"/>  </material>

  <material name="red">      <color rgba="1.00 0.00 0.00 1"/>  </material>
  <material name="green">    <color rgba="0.00 1.00 0.00 1"/>  </material>
  <material name="blue">     <color rgba="0.00 0.00 1.00 1"/>  </material>

  <material name="cyan">     <color rgba="0.00 1.00 1.00 1"/>  </material>
  <material name="magenta">  <color rgba="1.00 0.00 1.00 1"/>  </material>
  <material name="yellow">   <color rgba="1.00 1.00 0.00 1"/>  </material>

  <material name="orange">   <color rgba="1.00 0.65 0.00 1"/>  </material>
  
</robot>
